### DISCUSSION TITLE
Uninterpretable content in content stream when a PDF was created with text using PyMuPDF

### DISCUSSION THREAD

**Discussion by meghanaviyyapu** [pymupdf/PyMuPDF#1394] [Q&A] [answered]
I have created a PDF file and added some text using PyMuPDF library

import fitz
doc=fitz.open()
page=doc.newPage()
where=fitz.Point(200,10)
text="ABC"
page.insert_text(where,text)
path="F:\\PyMuPDF_test.pdf"
doc.save(path)

After opening the created PDF file and checking the contents using doc.xref_stream().decode(), I found [<414243>]. Can you let me know what these values are.
![image](https://user-images.githubusercontent.com/36301306/141422683-7c61eb9c-5654-4e71-b538-3e099b1243d7.png)

**Comment by JorjMcKie** [comment, marked as answer]
A typical "Discussions" item - no issue.

Content streams are written in PDF's mini-language by which the appearance of pages, annotations and some other object is defined.
The syntax is explained on pages 643 of https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdfs/PDF32000_2008.pdf.
Dozens of pages - too much to explain here. View this to be some type of source code of a programming language that you do not know.
On top, content streams are usually compressed, so you won't see that source code in ASCII.

To give you some start at least:

* **_q_**  - put current graphic state on the stack
* **_BT_** - begin a text object
* **_1 0 0 1 200 832 Tm_** - defines a text matrix. Numbers before Tm are the six matrix parameters. This one will position us 200 points from left and 832 points above bottom of the page.
* **_/helv 11 Tf_** - invoke text font named "helv" with a fontsize of 11
* **_[<414243>] TJ_** - output 3 characters in hex format, here "ABC"
* **_ET_** - end text object
* **_Q_** pull previous graphics state from stack

**Reply by meghanaviyyapu** [reply]
Thanks for the explanation. Can you let me know why in some cases there are special characters instead of characters in hex format in content stream.
![image](https://user-images.githubusercontent.com/36301306/141443710-6d591377-3d9e-4420-a5c3-bbdd1ec00556.png)
[sample-pdf-with-images.pdf](https://github.com/pymupdf/PyMuPDF/files/7526394/sample-pdf-with-images.pdf)

**Comment by JorjMcKie** [comment]
> Can you let me know why in some cases there are special characters instead of characters in hex format in content stream.

You are free to choose between hex and non-hex sometimes - depends on the font in use. For some fonts you cannot put the output characters ("ABC") in the contents.
Instead you must use the glyph number for each character and this must be given as an integer in hex format. The translation between the character's unicode and its glyph number is then taken over by the font's font program.

**Reply by meghanaviyyapu** [reply]
Does PyMuPDF have the capability of giving xref of text in a PDF.

**Comment by JorjMcKie** [comment]
> Does PyMuPDF have the capability of giving xref of text in a PDF.

Text in PDF does not have an XREF at all - nothing will be able to provide one.
Only the `/Contents` object has an xref - there may be several of those for a page.

**Comment by JorjMcKie** [comment]
What are you trying to achieve?
If you think you can reconstruct text by interpreting the contetns source yourself:  ... give up!
That's impossible ... or you rewrite a PDF viewer like Adobe Acrobat or MuPDF yourself.

**Reply by meghanaviyyapu** [reply]
I am trying to add tags to PDF 
![image](https://user-images.githubusercontent.com/36301306/141447368-e91d9fe7-84ed-4d42-a3d7-118db0246765.png)

For adding /K tag I need the text in the content stream so I was trying to find if there is any specific way to get a particular text from PDF.

**Reply by JorjMcKie** [reply]
I see. That will not work I am afraid - as I explained.

**Reply by meghanaviyyapu** [reply]
Yeah I understood. Do you have any idea on any python libraries that can be used for adding tags like this?

**Reply by JorjMcKie** [reply]
I am not aware of any, but maybe you find something [here](https://it.cornell.edu/accessibility/add-tags-pdf).
